gossip/src/duplicate_shred_handler.rs

```diff
@@ -144,18 +175,61 @@ impl DuplicateShredHandler {
             Some(SlotStatus::Frozen) => {
                 return false;
             }
-            Some(SlotStatus::UnfinishedProof(slot_map)) => {
-                if let Some(proof_chunkmap) = slot_map.peek(&data.from) {
-                    if proof_chunkmap.wallclock < data.wallclock {
-                        return false;
+            Some(SlotStatus::UnfinishedProof(slot_map)) => match slot_map.get(&data.from) {
+                None => {
+                    if slot_map.len() < MAX_PUBKEY_PER_SLOT {
+                        return true;
                     }
+                    return self.check_has_enough_stake_and_cleanup(&slot, &data.from);
                 }
-            }
+                Some(proof_chunkmap) => {
+                    return proof_chunkmap.wallclock >= data.wallclock;
+                }
```

Comment:
I don't want to put parts from different proofs together, even though they might be the same now, they might be different in the future.

And we define an order on wallclock so that our selection is kind of stable. I agree it's easy to attack, but I guess it's possible that a naive validator send out different proofs at different wallclock, then it's cleaner here to just stick to one of them.

Reply:
> I don't want to put parts from different proofs together

You don't have to do that. Just ignore it if there is already one, or replace the old one. It doesn't matter.

Reply:
Sure, I can do that, I don't think it's too important to make the order stable here.

Reply:
I still see this code using `wallclock` so many places.
I really don't think we should; just plz ignore if `wallclock` exist, and if you get two values just pick one; doesn't matter which.